examples/throughput/subscriber.cpp

```diff
@@ -0,0 +1,327 @@
+
+  return false;
+}
+
+template<typename T>
+class ThroughputListener: public dds::sub::DataReaderListener<T>
+{
+  public:
+  /*implementation of virtual functions*/
+
+  /*only on_data_available does anything*/
+  void on_data_available(dds::sub::DataReader<T>& rd) {
+    (void)do_take(rd);
+  }
+
+  /*all others are just dummies*/
+  void on_requested_deadline_missed(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::RequestedDeadlineMissedStatus&) { }
+
+  void on_requested_incompatible_qos(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::RequestedIncompatibleQosStatus&) { }
+
+  void on_sample_rejected(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::SampleRejectedStatus&) { }
+
+  void on_liveliness_changed(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::LivelinessChangedStatus&) { }
+
+  void on_subscription_matched(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::SubscriptionMatchedStatus&) { }
+
+  void on_sample_lost(
+      dds::sub::DataReader<T>&,
+      const dds::core::status::SampleLostStatus&) { }
+};
+
+int main (int argc, char **argv)
+{
+
+  if (parse_args(argc, argv) == EXIT_FAILURE)
+  {
+    return EXIT_FAILURE;
+  }
+
+  std::cout << subprefix << "Cycles: " << maxCycles << " | PollingDelay: " << pollingDelay.count() << " ms | Partition: " << partitionName << "\n"
+            << subprefix << "Using a " << (pollingDelay.count() > 0 ? "polling" : pollingDelay.count() < 0 ? "listener" : "waitset") << " approach.\n" << std::flush;
+
+  dds::domain::DomainParticipant participant(domain::default_id());
+
+  dds::topic::Topic<ThroughputModule::DataType> topic(participant, "Throughput");
+
+  dds::sub::qos::SubscriberQos sqos;
+  sqos.policy(dds::core::policy::Partition(partitionName));
+
+  dds::sub::Subscriber subscriber(participant, sqos);
+
+  dds::sub::qos::DataReaderQos rqos;
+  rqos.policy(
+    dds::core::policy::Reliability(
+      dds::core::policy::ReliabilityKind::Type::RELIABLE,
+      dds::core::Duration::from_secs(10)));
+  rqos.policy(
+    dds::core::policy::History(
+      dds::core::policy::HistoryKind::Type::KEEP_ALL,
+      0));
+  rqos.policy(
+    dds::core::policy::ResourceLimits(MAX_SAMPLES));
+
+  ThroughputListener<ThroughputModule::DataType> list;
+
+  dds::sub::DataReader<ThroughputModule::DataType>
+    reader(
+      subscriber,
+      topic,
+      rqos,
+      pollingDelay.count() < 0 ? &list : NULL,
+      pollingDelay.count() < 0 ? dds::core::status::StatusMask::data_available() : dds::core::status::StatusMask::none());
+
+  /* Process samples until Ctrl-C is pressed or until maxCycles */
+  /* has been reached (0 = infinite) */
+  signal (SIGINT, sigint);
+
+  if (wait_for_writer (reader))
+  {
+    std::cout << subprefix << "Waiting for samples...\n" << std::flush;
+
+    process_samples(reader, maxCycles);
+  }
+
+  return EXIT_SUCCESS;
+}
```

Comment:
Basically all the comments I gave for the Publisher regarding the use of command line parameters and the use of Qos also apply to the Subscriber.

Reply:
The choices in the publisher and subscriber regarding the QoS settings and the command line parameters were both done to make the C and C++ examples as interchangeable as possible.